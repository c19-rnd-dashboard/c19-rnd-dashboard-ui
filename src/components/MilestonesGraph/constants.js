export const status = {
  skipped: 'SKIPPED',
  completed: 'COMPLETED',
}

const discovery = 'discovery',
  preClinicalStudies = 'pre_clinical_studies',
  leadSelection = 'lead_selection',
  clinicalBatch = 'clinical_batch',
  ind = 'ind',
  phase1 = 'phase_1',
  phase2 = 'phase_2',
  phase3 = 'phase_3',
  nda = 'nda',
  onMarket = 'on_market'

export const phases = {
  discovery,
  preClinicalStudies,
  leadSelection,
  clinicalBatch,
  ind,
  phase1,
  phase2,
  phase3,
  nda,
  onMarket,
}

export const phaseDisplayName = {
  [discovery]: {
    label: 'Discovery',
    info:
      'Can include pre-formulation work, reverse vaccinology, antigen creation, pre-clinical ethical approvals',
  },
  [preClinicalStudies]: {
    label: 'Pre-Clinical Studies',
    info:
      'Can include in vitro and in vivo safety and efficacy studies, tox studies, animal challenge studies. Scientists have difficulty infecting mice with SARS-CoV-2 so this step might be skipped by certain companies',
  },
  [leadSelection]: {
    label: 'Lead Selection',
    info:
      'Lead product has been selected among potentially multiple candidates',
  },
  [clinicalBatch]: {
    label: 'Clinical Batch',
    info:
      'The actual vaccine material is manufactured in bulk to provide study drug for the clinical trials, this will need to go through pooling, formulation, sterile filtration, aseptic fill, packaging and labeling and then shipped',
  },
  [ind]: {
    label: 'IND or Equivalent Approval',
    info:
      'Sponsor received approval by the FDA for their Investigational New Drug appplication, allowing them to research an unapproved drug or study an approved drug for a new indication. There are equivalents in Europe (EMA), China (NMPA), Canada (Health Canada), Australia (TGA), UK (MHRA), India (CDSCO), Brazil (Health Ministry) and others. ',
  },
  [phase1]: {
    label: 'Phase 1',
    info:
      'Phase 1 studies are clinical trial in human subjects where the researchers are testing for the safety of the drug (as opposed to efficacy, i.e. if it works). These studies are usually run in healthy volunteers and might test multiple doses either simultaneously or in ascending order. In vaccine Phase 1 trial, researchers can also try to estimate efficacy by collecting data on antibody production.',
  },
  [phase2]: {
    label: 'Phase 2',
    info:
      'Phase 2 studies test fo the efficacy of the drug (i.e. if its works) in patients of the target patient population. Some studies combine Phase 1 and Phase 2 studies into one protocol, or combine Phase 2 and Phase 3 studies to expedite development, although there are typically still distinct stages that cover both phases in the protocol. ',
  },
  [phase3]: {
    label: 'Phase 3',
    info:
      'Phase 3 studies test efficacy and/or safety, but in a much wider population, ideally one that is demographically representative of the target population. The FDA typically requires two Phase 3 studies to be performed but it sometimes waves this requirement given the study drug particulars or past clinical data. ',
  },
  [nda]: {
    label: 'NDA',
    info:
      'In order for sponsors to provide their drugs on the market, they need to receive a New Drug Application approval from the FDA.  There are equivalents in Europe (EMA), China (NMPA), Canada (Health Canada), Australia (TGA), UK (MHRA), India (CDSCO), Brazil (Health Ministry) and others. ',
  },
  [onMarket]: {
    label: 'On The Market',
    info: '',
  },
}

export const phasesInOrder = [
  discovery,
  preClinicalStudies,
  leadSelection,
  clinicalBatch,
  ind,
  phase1,
  phase2,
  phase3,
  nda,
  onMarket,
]

export const phaseColor = {
  [discovery]: '#fff',
  [preClinicalStudies]: '#797979',
  [leadSelection]: '#941751',
  [clinicalBatch]: '#22528f',
  [ind]: '#3f8c27',
  [phase1]: '#8b551b',
  [phase2]: '#872110',
  [phase3]: '#eb51f7',
  [nda]: '#1431f5',
  [onMarket]: '#f09838',
}

export const timelinesEstimates = (() => {
  const partial = {
    optimistic: {
      [discovery]: 5,
      [preClinicalStudies]: 30,
      [leadSelection]: 5,
      [clinicalBatch]: 26,
      [ind]: 14,
      [phase1]: 75,
      [phase2]: 60,
      [phase3]: 120,
      [nda]: 21,
      [onMarket]: 30,
    },
    pesimistic: {
      [discovery]: 40,
      [preClinicalStudies]: 120,
      [leadSelection]: 25,
      [clinicalBatch]: 60,
      [ind]: 45,
      [phase1]: 180,
      [phase2]: 270,
      [phase3]: 365,
      [nda]: 60,
      [onMarket]: 180,
    },
  }
  const actual = Object.keys(partial.optimistic).reduce((acc, key) => {
    return {
      ...acc,
      [key]: (partial.optimistic[key] + partial.pesimistic[key]) / 2,
    }
  }, {})

  return {
    ...partial,
    actual,
  }
})()
