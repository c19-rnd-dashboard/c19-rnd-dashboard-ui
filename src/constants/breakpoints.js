export const smallWidth = 798

export const smallBreakpoint = `max-width: ${smallWidth}px`

export const bigBreakpoint = `min-width: ${smallWidth + 1}px`
