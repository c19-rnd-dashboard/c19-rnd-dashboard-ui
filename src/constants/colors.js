export const sadBlue = '#403FFF'
export const magenta = '#FF3FFF'
export const yellow = '#FEFF3F'
export const tourquese = '#00FEFF'
export const green = '#00FF00'

export const fontColor = 'white'
export const bg = 'black'
export const tileBg = '#222223'
