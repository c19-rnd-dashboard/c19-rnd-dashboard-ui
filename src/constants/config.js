export const mapboxApiKey =
  process.env.REACT_APP_MAPBOX_ACCESS_TOKEN ||
  process.env.STORYBOOK_MAPBOX_ACCESS_TOKEN

export const apiUrl =
  process.env.REACT_APP_API_URL ||
  process.env.STORYBOOK_API_URL ||
  'https://c19-vac-rnd-dash-staging-api.herokuapp.com'

export const faqUrl =
  process.env.REACT_APP_FAQ_URL ||
  'https://docs.google.com/document/d/1Mw-CfAzezFh6sFKRKlERU2HxX_e5Lu5bXSUcjVD5gso'

export const howYouCanHelpUrl = process.env.REACT_APP_HOW_YOU_CAN_HELP || '#'

export const spreadsheetDataSource =
  process.env.REACT_APP_SPREADSHEET_DATASOURCE || '#'

export const contactUsUrl =
  process.env.REACT_APP_CONTACT_US_URL || 'mailto:info@coviddash.org'
