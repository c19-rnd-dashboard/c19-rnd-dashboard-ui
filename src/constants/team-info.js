import JonSanders from '../assets/images/jonsanders.jpg'
import Mats from '../assets/images/MatsOlsen.jpg'
import Liam from '../assets/images/liam_headshot.jpg'
import Nestor from '../assets/images/nestor.png'
import Nafi from '../assets/images/Nafi.jpg'
import Sunny from '../assets/images/sunny.png'
import Vinny from '../assets/images/vinny.jpg'

export default [
  {
    name: 'Mats Olson',
    title: 'Founder',
    blurb: [
      'Founded COVID Dash on March 17, 2020.',
      '13 years clinical trial industry experience.',
      'SVP of clinical trial technology solution provider, Prevail InfoWork.',
    ],
    link: { title: 'LinkedIn', url: 'https://www.linkedin.com/in/mats-olsen/' },
    image: Mats,
  },
  {
    name: 'Nestor Torbon',
    title: 'Software Developer',
    blurb: ['Software developer at Splunk.'],
    link: { title: 'Github', url: 'https://github.com/Nestyko' },
    image: Nestor,
  },
  {
    name: 'Jonathan Sanders',
    title: 'Software Developer',
    blurb: ['Fullstack software developer at Codecademy.'],
    link: { title: 'Github', url: 'https://github.com/j10sanders' },
    image: JonSanders,
  },
  {
    name: 'Liam Rosen',
    title: 'Marketing and Operations',
    blurb: [
      'Prior to COVID Trial Volunteer Dash, Liam helped start GetUsPPE, the largest US nonprofit helping link donors of personal protective equipment to facilities in need. Outside of nonprofit work, he is the CEO of TopScore, a sports technology company.',
    ],
    link: { title: 'Website', url: 'https://liamrosen.com/' },
    image: Liam,
  },
  {
    name: 'Vincent Brandon',
    title: 'Data Engineer',
    blurb: [
      'PMP certified project manager.  Data scientist for the State of Utah.  Past biochemist and pharmaceutical trials coordinator.',
    ],
    image: Vinny,
  },
  {
    name: 'Nafi Adibi',
    title: 'UX Designer',
    blurb: [''],
    link: { title: 'LinkedIn', url: 'https://www.linkedin.com/in/nafiadibi' },
    image: Nafi,
  },
  {
    name: 'Sunny Su',
    title: 'UX Designer',
    blurb: [''],
    link: { title: 'LinkedIn', url: 'https://www.linkedin.com/in/sunnylsu/' },
    image: Sunny,
  },
]
