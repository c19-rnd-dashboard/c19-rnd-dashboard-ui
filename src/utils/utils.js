import moment from 'moment'
import { map, pipe } from 'sanctuary'

export const log = tag => value => {
  console.log({ tag, value })
  return value
}

export const mapTwoAtTime = (fn, xs) => {
  const [head, subHead, ...rest] = xs
  if (!head) {
    return []
  }
  const [first, second] = fn(head, subHead)
  return [first, ...mapTwoAtTime(fn, [second, ...rest])]
}

export const isVaccine = ({ interventionType = '' }) =>
  interventionType.toLowerCase().includes('vaccine')

export const addToDate = (date, duration, timeUnit = 'days') =>
  moment(date).add(duration, timeUnit).toISOString()

export const isValidDate = date => moment(date).isValid()

export const count = ls =>
  ls.reduce(
    (acc, x) =>
      !Object.prototype.hasOwnProperty.call(acc, x)
        ? { ...acc, [x]: 1 }
        : { ...acc, [x]: acc[x] + 1 },
    {}
  )

export const countBy = fn => pipe([map(fn), count])

export const converCountIntoChartData = pipe([
  Object.entries,
  map(([key, value]) => ({
    id: key,
    value: value,
  })),
])
