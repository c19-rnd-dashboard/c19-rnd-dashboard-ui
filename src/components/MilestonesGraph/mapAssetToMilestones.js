import moment from 'moment'
import { pipe } from 'sanctuary'
import { status, phasesInOrder, timelinesEstimates } from './constants'
import { mapTwoAtTime, isVaccine, addToDate } from 'utils/utils'

const { skipped } = status

export const filterAndSortMilestones = milestones =>
  [...milestones]
    .filter(({ status }) => status !== skipped)
    .sort(
      ({ name: nameA }, { name: nameB }) =>
        phasesInOrder.indexOf(nameA) - phasesInOrder.indexOf(nameB)
    )

const addDurationToMilestones = ({ now, delta }) => milestones => {
  let buffer = []
  for (let i = 1; i <= milestones.length; i++) {
    const first = milestones[i - 1]
    if (first.duration) {
      continue
    }
    const second = milestones[i] ? milestones[i] : { date: now }
    const { date: lastDate } = first
    const { date: newDate } = second
    const start = moment(lastDate)
    const end = moment(newDate)
    const duration = end.diff(start, 'days') || delta[first.name]
    buffer = buffer.concat({ ...first, duration, start, end })
  }
  return buffer
}

const addPercentageToMilestones = ms => {
  const totals = ms.map(({ values }) =>
    values.reduce((acc, { duration = 0 }) => duration + acc, 0)
  )
  const [max] = totals.slice().sort((a, b) => b - a)
  return ms.map(m => ({
    ...m,
    values: m.values.map(x => ({
      ...x,
      percentage: Math.round(((x.duration * 100) / max) * 100) / 100,
    })),
  }))
}

const mapResultToMilestoneStructure = ms =>
  ms.map(m => ({
    ...m,
    values: m.values.map(({ name, percentage, start, end }) => ({
      className: name,
      name,
      value: `${percentage}%`,
      start: moment(start).toISOString(),
      end: moment(end).toISOString(),
    })),
  }))

const getAdditionalPhases = phase => {
  if (!phase) {
    return phasesInOrder
  }
  const index = phasesInOrder.indexOf(phase)
  return phasesInOrder.filter(p => phasesInOrder.indexOf(p) > index)
}

const estimateFutureDates = (ms, delta) => {
  return mapTwoAtTime((first, second) => {
    const duration = delta[first.name]
    const newEnd = addToDate(first.start, duration, 'days')
    const newFirst = { ...first, end: newEnd, duration }
    const newSecond = { ...second, start: newEnd }
    return second ? [newFirst, newSecond] : [newFirst]
  }, ms)
}

const transformWithDurations = ({ now, delta }) =>
  pipe([filterAndSortMilestones, addDurationToMilestones({ now, delta })])

const t2 = pipe([addPercentageToMilestones, mapResultToMilestoneStructure])

const getEstimationCheckingItIsNotGtActual = (delta, actual) => {
  const duration = delta[actual.name]
  const newEnd = addToDate(actual.start, duration, 'days')
  const isLessThanActual = new Date(newEnd) < new Date(actual.end)
  return isLessThanActual ? { ...delta, [actual.name]: actual.duration } : delta
}

export const mapAssetToMilestones = now => ({
  milestones,
  interventionType,
}) => {
  if (!milestones) {
    return []
  }
  const type = isVaccine({ interventionType }) ? 'vaccine' : 'treatment'

  const actualMilestonesWithDuration = transformWithDurations({
    now,
    delta: timelinesEstimates.actual,
  })(milestones)
  const [latestKnownMilestone] = actualMilestonesWithDuration.slice(-1)
  const estimationPhases = getAdditionalPhases(
    latestKnownMilestone.name
  ).map(phase => ({ name: phase }))
  const optimisticEstimations = estimateFutureDates(
    [latestKnownMilestone].concat(estimationPhases),
    getEstimationCheckingItIsNotGtActual(
      timelinesEstimates.optimistic,
      latestKnownMilestone
    )
  )
  const pessimisticEstimations = estimateFutureDates(
    [latestKnownMilestone].concat(estimationPhases),
    getEstimationCheckingItIsNotGtActual(
      timelinesEstimates.pessimistic,
      latestKnownMilestone
    )
  )
  const result = [
    {
      name: 'Actual',
      values: actualMilestonesWithDuration,
    },
  ]
  return t2(
    type === 'vaccine'
      ? [
        {
          name: 'Optimistic',
          values: actualMilestonesWithDuration
            .slice(0, -1)
            .concat(optimisticEstimations),
        },
        {
          name: 'Pessimistic',
          values: actualMilestonesWithDuration
            .slice(0, -1)
            .concat(pessimisticEstimations),
        },
        ...result,
      ]
      : result
  )
}
