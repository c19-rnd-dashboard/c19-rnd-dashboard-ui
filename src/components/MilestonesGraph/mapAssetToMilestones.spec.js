import asset from './mocks/asset.json'
import assets from '../../mocks/assets.json'
import { mapAssetToMilestones } from './mapAssetToMilestones'
import moment from 'moment-timezone'
import { profileTime } from 'utils/utils'
import { status } from './constants'
const { skipped } = status
moment.tz.setDefault('UTC')

const now = moment('2020-05-09').toISOString()

it('should not take too long', () => {
  const assetsWithMilestones = assets.filter(
    ({ milestones = [] }) =>
      milestones.filter(({ status }) => status !== skipped).length > 0
  )
  const { duration } = profileTime(() =>
    assetsWithMilestones.map(asset => mapAssetToMilestones(now)(asset))
  )
  expect(duration).toBeLessThan(2200)
})

it('should return the expected result', () => {
  expect(mapAssetToMilestones(now)(asset)).toMatchSnapshot()
})
