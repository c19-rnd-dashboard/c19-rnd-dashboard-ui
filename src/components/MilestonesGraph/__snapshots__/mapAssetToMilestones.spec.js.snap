// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the expected result 1`] = `
Array [
  Object {
    "name": "Optimistic",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.18%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.23%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.32%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.07%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-30T01:00:00.000Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "6.7%",
      },
      Object {
        "className": "phase_2",
        "end": "2020-10-27T00:00:00.000Z",
        "start": "2020-05-30T01:00:00.000Z",
        "value": "13.39%",
      },
      Object {
        "className": "phase_3",
        "end": "2021-06-24T01:00:00.000Z",
        "start": "2020-10-27T00:00:00.000Z",
        "value": "21.43%",
      },
      Object {
        "className": "nda",
        "end": "2021-07-15T01:00:00.000Z",
        "start": "2021-06-24T01:00:00.000Z",
        "value": "1.88%",
      },
      Object {
        "className": "on_market",
        "end": "2021-08-14T01:00:00.000Z",
        "start": "2021-07-15T01:00:00.000Z",
        "value": "2.68%",
      },
    ],
  },
  Object {
    "name": "Pesimistic",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.18%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.23%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.32%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.07%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-09-12T00:00:00.000Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "16.07%",
      },
      Object {
        "className": "phase_2",
        "end": "2021-06-09T01:00:00.000Z",
        "start": "2020-09-12T00:00:00.000Z",
        "value": "24.11%",
      },
      Object {
        "className": "phase_3",
        "end": "2022-06-09T01:00:00.000Z",
        "start": "2021-06-09T01:00:00.000Z",
        "value": "32.59%",
      },
      Object {
        "className": "nda",
        "end": "2022-08-08T01:00:00.000Z",
        "start": "2022-06-09T01:00:00.000Z",
        "value": "5.36%",
      },
      Object {
        "className": "on_market",
        "end": "2023-02-04T00:00:00.000Z",
        "start": "2022-08-08T01:00:00.000Z",
        "value": "16.07%",
      },
    ],
  },
  Object {
    "name": "Actual",
    "values": Array [
      Object {
        "className": "discovery",
        "end": "2020-01-13T00:00:00.000Z",
        "start": "2020-01-11T00:00:00.000Z",
        "value": "0.18%",
      },
      Object {
        "className": "lead_selection",
        "end": "2020-02-07T00:00:00.000Z",
        "start": "2020-01-13T00:00:00.000Z",
        "value": "2.23%",
      },
      Object {
        "className": "clinical_batch",
        "end": "2020-03-04T00:00:00.000Z",
        "start": "2020-02-07T00:00:00.000Z",
        "value": "2.32%",
      },
      Object {
        "className": "ind",
        "end": "2020-03-16T00:00:00.000Z",
        "start": "2020-03-04T00:00:00.000Z",
        "value": "1.07%",
      },
      Object {
        "className": "phase_1",
        "end": "2020-05-09T00:00:00.000Z",
        "start": "2020-03-16T00:00:00.000Z",
        "value": "4.73%",
      },
    ],
  },
]
`;
